controllers/tykoasapidefinition_controller.go

```diff
@@ -0,0 +1,274 @@
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the TykOASApiDefinition object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *TykOASApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("tykoasapidefinition", req.NamespacedName)
+
+	log.Info("Reconciling TykOASApiDefinition")
+
+	tykOASDef := &tykv1alpha1.TykOASApiDefinition{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, tykOASDef)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	_, ctx, err = HttpContext(ctx, r.Client, r.Env, tykOASDef, log)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	_, err = util.CreateOrUpdate(ctx, r.Client, tykOASDef, func() error {
+		if !tykOASDef.ObjectMeta.DeletionTimestamp.IsZero() {
+			err := r.delete(ctx, tykOASDef)
+			if err != nil {
+				return err
+			}
+
+			util.RemoveFinalizer(tykOASDef, keys.TykOASApiDefinitionFinalizerName)
+
+			return nil
+		}
+
+		util.AddFinalizer(tykOASDef, keys.TykOASApiDefinitionFinalizerName)
+
+		id, err := r.createOrUpdateTykOAS(ctx, tykOASDef)
+		if err != nil {
+			return err
+		}
+
+		if tykOASDef.Status.ApiID == "" {
+			tykOASDef.Status.ApiID = id
+
+			err = r.Client.Status().Update(ctx, tykOASDef)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		log.Error(err, "Failed to create/update TykOASApiDefinition")
+		return ctrl.Result{RequeueAfter: queueAfter}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *TykOASApiDefinitionReconciler) createOrUpdateTykOAS(
+	ctx context.Context,
+	tykOASDef *tykv1alpha1.TykOASApiDefinition,
+) (string, error) {
+	r.Log.Info("Creating/Updating OAS API on Tyk")
+
+	var cm v1.ConfigMap
+
+	cmName := types.NamespacedName{Name: tykOASDef.Spec.OASRef.Name, Namespace: tykOASDef.Spec.OASRef.Namespace}
+
+	err := r.Client.Get(ctx, cmName, &cm)
+	if err != nil {
+		return "", err
+	}
+
+	data := cm.Data[tykOASDef.Spec.OASRef.KeyName]
+	if data == "" {
+		err = errors.New("OAS Spec is empty")
+		return "", err
+	}
+
+	id := tykOASDef.Status.ApiID
+	if id == "" {
+		var result *model.Result
+
+		result, err = klient.Universal.OAS().Create(ctx, data)
+
+		id = result.Meta
+	} else {
+		err = klient.Universal.OAS().Update(ctx, id, data)
+	}
+
+	if err != nil {
+		r.Log.Error(err, "Failed to create/update OAS API Definition")
+		return "", err
+	}
+
+	r.Log.Info("Successfully created/updated OAS API on Tyk", "id", id)
```

Comment:
maybe, we can print logs after sending Update or Create requests.
So, for example, after creating the OAS, we can print `successfully created OAS on Tyk`.